4-Shaders-Uniforms/Window.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System;
+using System.Diagnostics;
+using OpenTK;
+using OpenTK.Graphics;
+using OpenTK.Graphics.OpenGL4;
+using OpenTK.Input;
+using LearnOpenTK.Common;
+
+namespace LearnOpenTK
+{
+    // This project will explore how to use uniform variable type which allows you to assign values
+    // To Shaders at any point during the project
+    public class Window : GameWindow
+    {
+
+        private readonly float[] _vertices =
+        {
+            -0.5f, -0.5f, 0.0f, // Bottom-left vertex
+             0.5f, -0.5f, 0.0f, // Bottom-right vertex
+             0.0f,  0.5f, 0.0f  // Top vertex
+        };
+
+
+        // So we're going make the triangle pulsate between a color range
+        // And in order to do that we'll need a constantly changing value
+        // The stopwatch is perfect for this as it's a constantly going up
+        private Stopwatch _timer;
+
+        private int _vertexBufferObject;
+
+        private int _vertexArrayObject;
+
+        private Shader _shader;
+
+        public Window(int width, int height, string title)
+            : base(width, height, GraphicsMode.Default, title)
+        {
+        }
+
+        protected override void OnLoad(EventArgs e)
+        {
+            GL.ClearColor(0.2f, 0.3f, 0.3f, 1.0f);
+
+            _vertexBufferObject = GL.GenBuffer();
+
+            GL.BindBuffer(BufferTarget.ArrayBuffer, _vertexBufferObject);
+            GL.BufferData(BufferTarget.ArrayBuffer, _vertices.Length * sizeof(float), _vertices, BufferUsageHint.StaticDraw);
+
+            _shader = new Shader("Shaders/shader.vert", "Shaders/shader.frag");
+
+            _shader.Use();
+
+            _vertexArrayObject = GL.GenVertexArray();
+            GL.BindVertexArray(_vertexArrayObject);
+
+            GL.VertexAttribPointer(0, 3, VertexAttribPointerType.Float, false, 3 * sizeof(float), 0);
+            GL.EnableVertexAttribArray(0);
+
+            GL.BindBuffer(BufferTarget.ArrayBuffer, _vertexBufferObject);
+
+            int nrAttributes = 0;
+            GL.GetInteger(GetPName.MaxVertexAttribs, out nrAttributes);
+            Console.WriteLine("Maximum number of vertex attributes supported: " + nrAttributes);
+
+            // We start the stop watch here as this method is only called once 
+            _timer = new Stopwatch();
+            _timer.Start();
+
+            base.OnLoad(e);
+        }
+
+        protected override void OnRenderFrame(FrameEventArgs e)
+        {
+            GL.Clear(ClearBufferMask.ColorBufferBit);
+
+            _shader.Use();
+
+            // So here we get the total seconds that have elapsed since the last time this method has been called
+            // And we assign it to the timeValue variable so it can be used for the pulsating color 
```

Comment:
not too important, but the stopwatch elapsed actually gives the time since the last time stopwatch.reset was called.
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch?view=netcore-3.1